### DISCUSSION TITLE
Sate chart , how to get list of states ,

### DISCUSSION THREAD

**Discussion by chansdad** [statelyai/xstate#2655] [Q&A]
Hello All  , New to xstate here. Great product so far based on what i am reading.  

My requirement is to  interpret a state chart and gather all the list of possible states including initial and final states , machine id ,  events, all possible transitions and actions , guarded transitions etc that a state chart has in its state chart definition . 

My goal is to take execute a state chart and be able to expose the states and events as a service to a different application so that other application can interact with the state chart .  In its simplest form , if i have following counter state chart  i want to be able to start this state chart with an id assigned dynamically and then let it run , and have another application interact with it , ex , other application will invoke the machine service , ex : http://server/countermachine/1/inc , so that the counter for state machine with id is incremented.  Note that In my use case , machine model is not known in advance and a user will load his or her own state chart as such i would like to know what API i can use and in a way i think the visualize app uses some API to display the states, events , and transitions etc . Appreciate if any one can point me to the right docs .

```
const counterMachine = createMachine({
  initial: 'active',
  context: {
    count: 0
  },
  states: {
    active: {
      on: {
        INC: { actions: assign({ count: increment }) },
        DEC: { actions: assign({ count: decrement }) }
      }
    }
  }
});


```

**Comment by Andarist** [comment]
I'm unsure if I got the use case right but it sounds like you might want to do some custom processing here. It all depends on what exact information do you need - since our configs are just objects you should be able to traverse that object however you want. You can also try to call `JSON.stringify` on the machine and crawl the resulting, more normalized, format.

**Reply by chansdad** [reply]
So here is my use case , I want the user to be able to upload a state chart and want to expose that state machine as a service . which means i have no idea what state chart he will be uploading. hence wanted to know if there is an API , where in I can execute the state chart and gather all the potential states , events, transitions etc .If i have to crawl ,not sure if xstate has a defined schema , users may load a xstate machine which is incomplete or not following the spec .. Is crawling the object only option for me?

**Reply by davidkpiano** [reply]
> Is crawling the object only option for me?

Right now, yes.

We will probably be adding helper functions and more insightful properties to `@xstate/graph` though, for this use-case and others.

**Comment by pke** [comment]
I created a similar such service 2 years ago. It exposed a bunch of (server persitable) state charts as SIREN responses.
I used the `meta` info on the machine and on the states to describe the fields. The allowed events can be returned to the client via `service.state.nextEvents`.

Nowadays I would combine the power of TypeScript and something like `joi`. Since TS is compile time only but we need runtime information too so that the generic client knows what events are allowed and which inputs they need to work.

I haven't checked but I hope `joi` would work together with TypeScript and then since we know which events are allowed at any given state (internal events are not exposed to clients!) we can then just use the `joi` definitions of each event to return to the client.

This could be really powerful, because it could contain all kind of constrains like min and max string length or accepted patterns that the client could already validate (if it understands them).

My SIREN responses were interpreted by a client side react app as well as a command line client for SIREN. Both could traverse the state charts flawlessly for what I had back then.

My poster example for that service was a electric kettle state chart, which even made use of Server Side Events to pro-actively update the client, when the temperature had changed (over time).

```ts
const { Machine, assign } = require("xstate")

const HEAT_CAPACITY_OF_WATER = 4180

const heatingDuration = ({
  power,
  efficiency,
  amount,
  waterTemp,
  targetTemp = 100
}) => {
  const effectivePower = power * (efficiency * 0.01)
  const heat = HEAT_CAPACITY_OF_WATER * amount * (targetTemp - waterTemp)
  return Math.round(heat / effectivePower)
}

const roundToTwoDigitsAfterComma = number => (
  parseFloat((Math.round(number * 100) / 100).toFixed(2))
)

const kettle = Machine({
  id: "kettle",
  initial: "idle",
  meta: {
    title: "Kettle"
  },
  context: {
    capacity: 1600, // in millilitre
    power: 2000, // in Watts
    efficiency: 90, // in %
    amount: 0, // in millilitre
    temp: 0, // in °C
    targetTemp: 100, // °C
  },
  states: {
    lidOpen: {
      on: {
        FILL_WATER: {
          target: "idle",
          actions: "fillWater",
          cond: "validFillEvent",
        },
        CLOSE_LID: "idle"
      },
      meta: {
        on: {
          FILL_WATER: {
            title: "Pour water into the kettle",
            fields: [
              {
                name: "amount",
                type: "number",
                min: 1,
                max: "maxFillAmount"
              },
              {
                name: "temp",
                type: "number",
                max: 100,
              }
            ]
          }
        }
      }
    },
    idle: {
      on: {
        OPEN_LID: "lidOpen",
        START: {
          target: "heating",
          actions: "setStartTime",
          cond: "notEmpty"
        },
        POUR_WATER: {
          actions: "pourWater",
          cond: "canPourWater"
        }
      },
      meta: {
        on: {
          POUR_WATER: {
            title: "Pour water out of the kettle",
            fields: [
              {
                name: "amount",
                type: "number",
                min: 1,
                max: "maxPourAmount",
                value: "maxPourAmount"
              }
            ]
          }
        }
      }
    },
    heating: {
      after: {
        NEXT_TEMP_TIME: {
          target: "heating",
          actions: "incTemp"
        }
      },
      on: {
        "": {
          target: "idle",
          cond: "targetTempReached"
        },
        DISCONNECT: "idle",
        LIFT_UP: "idle",
        STOP: "idle"
      }
    }
  }
}, {
  meta: {
    values: {
      maxFillAmount: context => context.capacity - context.amount,
      maxPourAmount: context => context.amount
    }
  },
  actions: {
    fillWater: assign({
      amount: (context, event) => context.amount + (0 + event.amount),
      temp: (context, event) => (
        roundToTwoDigitsAfterComma((context.amount * context.temp + Number(event.amount) * Number(event.temp)) / (context.amount + Number(event.amount)))
      )
    }),
    pourWater: assign({
      amount: (context, event) => context.amount - event.amount,
    }),
    setStartTime: assign({ startTime: () => Date.now() }),
    incTemp: assign({ temp: context => context.temp + 1 })
  },
  delays: {
    NEXT_TEMP_TIME: (context) => {
      const time = heatingDuration({
        power: context.power,
        efficiency: context.efficiency,
        waterTemp: context.temp,
        amount: context.amount,
        targetTemp: context.temp + 1
      })
      return time
    }
  },
  guards: {
    validFillEvent: (context, event) => (
      context.amount < context.capacity
      && event.amount > 0
      && event.temp !== undefined
      && context.amount + event.amount <= context.capacity
    ),
    canPourWater: (context, event) => (
      context.amount > 0
      && event.amount <= context.amount
    ),
    notEmpty: (context) => context.amount > 0,
    targetTempReached: (context) => {
      return context.temp === context.targetTemp
    }
  }
})

module.exports = kettle
```

And thats how I composed the SIREN response:

```ts
function toSiren(service, annotate = {}) {
  const current = service.state
  return {
    class: service.id,
    title: current.meta.title || (service.machine.meta && service.machine.meta.title) || current.id,
    properties: current.context,
    links: [
      {
        rel: ["self"],
        href: annotate.href("")
      }
    ],
    entities: [
      {
        class: "state",
        properties: current
      }
    ],
    actions: current.nextEvents.reduce((actions, event) => {
      let meta = current.meta[`${service.id}.${current.toStrings()[0]}`] 
      if (!meta) {
        meta =  { on: { [event]: {} } }
      }
      const { fields, ...rest} = ((meta.on||{})[event] || [])
      if (Array.isArray(fields) || service.machine.transition(current, event).changed) {
        let action = {
          name: event,
          href: annotate.href && annotate.href(event),
          method: "PUT",
          fields: (fields||[]).map(field => {
            return Object.keys(field).reduce((field, key) => {
              const resolver = ((service.machine.options.meta||{}).values||{})[field[key]]
              if (resolver) {
                if (typeof resolver === "function") {
                  field[key] = resolver(current.context)
                }
              }
              return field
            }, field)
          }),
          ...rest
        }
        actions.push(action)
      }
      return actions
    }, [])
  }
} 
```

**Reply by chansdad** [reply]
Thank you very much for sharing this , quite helpful .
